File: frontend/src/api/marketHealth.ts
export interface MarketHealthCsvResolution {
  valid_rows: number;
  skipped_rows: number;
  duplicate_rows: number;
  malformed_rows: number;
}

export interface MarketHealthProviderSkip {
  symbol: string;
  message: string;
}

export interface MarketHealthUniverse {
  version: number;
  file_name: string;
  symbols: string[];
  imported_count: number;
  usable_count: number;
  csv_resolution: MarketHealthCsvResolution;
  provider_skips: {
    finviz: MarketHealthProviderSkip[];
    yahoo: MarketHealthProviderSkip[];
  };
  created_at: string;
}

export interface MarketHealthJobSnapshot {
  revision: number;
  job_id?: number;
  phase: "no_universe" | "parsing" | "stale" | "running" | "pausing" | "paused" | "ready" | "failed";
  progress?: MarketHealthProgress | null;
}

export type MarketHealthProgressStepState = "pending" | "running" | "completed" | "failed";
export type MarketHealthTickerState = "pending" | "current" | "completed" | "skipped" | "failed";

export interface MarketHealthProgressStep {
  state: MarketHealthProgressStepState;
  total: number;
  completed: number;
  skipped: number;
  failed: number;
  current_symbol: string | null;
  processed_symbols: string[];
  message: string | null;
  elapsed_seconds: number;
}

export interface MarketHealthTickerProgress {
  symbol: string;
  state: MarketHealthTickerState;
  message: string | null;
  benchmark: boolean;
}

export interface MarketHealthProgress {
  completed_work_items: number;
  total_work_items: number;
  completed_tickers: number;
  total_tickers: number;
  cached_count: number;
  refreshed_count: number;
  failed_count: number;
  provider_skips: MarketHealthUniverse["provider_skips"];
  ticker_statuses: MarketHealthTickerProgress[];
  finviz: MarketHealthProgressStep;
  yahoo: MarketHealthProgressStep;
}

export interface MarketHealthPoint { date: string; value: number | null; matching_count: number; valid_count: number }
export interface MarketHealthSeries {
  name: string;
  points: MarketHealthPoint[];
  summary: {
    current: number | null;
    change_5d: number | null;
    change_20d: number | null;
    matching_count: number | null;
    valid_count: number | null;
  };
}
export interface MarketHealthChart { title: string; percent: boolean; series: MarketHealthSeries[] }
export interface MarketHealthGroup {
  key: string; name: string; member_count: number; eligible_count: number; small_group: boolean;
  above_sma20_valid_count: number; above_sma50_valid_count: number; new_high_valid_count: number;
  new_low_valid_count: number; outperform_20_valid_count: number; outperform_63_valid_count: number;
  above_sma20_percent: number | null; above_sma50_percent: number | null;
  new_high_percent: number | null; new_low_percent: number | null;
  outperform_20_percent: number | null; outperform_63_percent: number | null;
  above_sma50_change_5d: number | null; above_sma50_change_20d: number | null;
}
export interface MarketHealthLeadingStock {
  symbol: string;
  return_20: number; return_selected: number; excess_20: number; excess_selected: number;
  above_sma20: boolean; above_sma50: boolean | null; distance_from_high_63: number | null;
  new_high_63: boolean | null; adv20: number;
  industry_key: string | null;
  industry_group: string | null;
  themes: string[];
}
export interface MarketHealthTabResponse {
  tab: string;
  benchmark: string;
  latest_session: string;
  charts: MarketHealthChart[];
  groups: MarketHealthGroup[];
  leading_stocks: MarketHealthLeadingStock[];
  selected_group: string | null;
  leader_sessions: number;
  eligible_count: number;
  universe_count: number;
  group_members: string[];
}

export async function fetchMarketHealthUniverse(
  signal?: AbortSignal,
): Promise<MarketHealthUniverse | null> {
  const response = await fetch("/api/market-health/universe", { signal });
  if (response.status === 204) return null;
  if (!response.ok) throw new Error(await responseError(response, "Failed to load Market Health universe"));
  return response.json() as Promise<MarketHealthUniverse>;
}

export async function uploadMarketHealthUniverse(file: File): Promise<MarketHealthUniverse> {
  const form = new FormData();
  form.append("files", file, file.name);
  const response = await fetch("/api/market-health/universe", { method: "POST", body: form });
  if (!response.ok) throw new Error(await responseError(response, "Failed to upload Market Health universe"));
  return response.json() as Promise<MarketHealthUniverse>;
}

export async function changeMarketHealthLifecycle(action: "pause" | "resume"): Promise<MarketHealthJobSnapshot> {
  const response = await fetch(`/api/market-health/${action}`, { method: "POST" });
  if (!response.ok) throw new Error(await responseError(response, `Failed to ${action} Market Health job`));
  return response.json() as Promise<MarketHealthJobSnapshot>;
}

export async function restartMarketHealth(action: "refresh" | "retry"): Promise<MarketHealthJobSnapshot> {
  const response = await fetch(`/api/market-health/${action}`, { method: "POST" });
  if (!response.ok) throw new Error(await responseError(response, `Failed to ${action} Market Health job`));
  return response.json() as Promise<MarketHealthJobSnapshot>;
}

export async function fetchMarketHealthTab(tab: string, group?: string, signal?: AbortSignal, leaderSessions = 63): Promise<MarketHealthTabResponse> {
  const params = new URLSearchParams({ tab });
  if (group !== undefined) params.set("group", group);
  if (tab === "leading_stocks") params.set("leader_sessions", String(leaderSessions));
  const response = await fetch(`/api/market-health/tab?${params}`, { signal });
  if (!response.ok) throw new Error(await responseError(response, "Failed to calculate Market Health tab"));
  return response.json() as Promise<MarketHealthTabResponse>;
}

async function responseError(response: Response, fallback: string) {
  try {
    const body = await response.json() as { error?: string };
    return body.error ?? `${fallback}: HTTP ${response.status}`;
  } catch {
    return `${fallback}: HTTP ${response.status}`;
  }
}

File: src/models/market_health.rs
use chrono::{DateTime, NaiveDate, Utc};
use serde::Serialize;

use super::TickerSymbol;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MarketHealthPhase {
    NoUniverse,
    Parsing,
    Stale,
    Running,
    Pausing,
    Paused,
    Ready,
    Failed,
}

impl MarketHealthPhase {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::NoUniverse => "no_universe",
            Self::Parsing => "parsing",
            Self::Stale => "stale",
            Self::Running => "running",
            Self::Pausing => "pausing",
            Self::Paused => "paused",
            Self::Ready => "ready",
            Self::Failed => "failed",
        }
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthJobSnapshot {
    pub revision: u64,
    pub job_id: Option<u64>,
    pub phase: MarketHealthPhase,
    pub work_plan: Option<MarketHealthWorkPlan>,
    pub progress: Option<MarketHealthPreparationProgress>,
}

impl Default for MarketHealthJobSnapshot {
    fn default() -> Self {
        Self {
            revision: 0,
            job_id: None,
            phase: MarketHealthPhase::NoUniverse,
            work_plan: None,
            progress: None,
        }
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthPreparationProgress {
    pub completed_work_items: usize,
    pub total_work_items: usize,
    pub completed_tickers: usize,
    pub total_tickers: usize,
    pub cached_count: usize,
    pub refreshed_count: usize,
    pub failed_count: usize,
    pub provider_skips: MarketHealthProviderSkips,
    pub ticker_statuses: Vec<MarketHealthTickerProgress>,
    pub finviz: MarketHealthProviderStepProgress,
    pub yahoo: MarketHealthProviderStepProgress,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthTickerProgress {
    pub symbol: TickerSymbol,
    pub state: MarketHealthTickerState,
    pub message: Option<String>,
    pub benchmark: bool,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MarketHealthTickerState {
    Pending,
    Current,
    Completed,
    Skipped,
    Failed,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthProviderStepProgress {
    pub state: MarketHealthProviderStepState,
    pub total: usize,
    pub completed: usize,
    pub skipped: usize,
    pub failed: usize,
    pub current_symbol: Option<TickerSymbol>,
    pub processed_symbols: Vec<TickerSymbol>,
    pub message: Option<String>,
    pub elapsed_seconds: u64,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MarketHealthProviderStepState {
    Pending,
    Running,
    Completed,
    Failed,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthWorkPlan {
    pub range: MarketHealthSessionRange,
    pub ticker_count: usize,
    pub cached_count: usize,
    pub work_items: Vec<MarketHealthWorkItem>,
    pub benchmark: MarketHealthBenchmarkWork,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthBenchmarkWork {
    pub symbol: TickerSymbol,
    pub needs_yahoo: bool,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthSessionRange {
    pub source_start: NaiveDate,
    pub display_start: NaiveDate,
    pub latest_session: NaiveDate,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthWorkItem {
    pub symbol: TickerSymbol,
    pub needs_finviz: bool,
    pub needs_yahoo: bool,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthUniverse {
    pub version: u8,
    pub file_name: String,
    pub symbols: Vec<TickerSymbol>,
    pub imported_count: usize,
    pub usable_count: usize,
    pub csv_resolution: MarketHealthCsvResolution,
    pub provider_skips: MarketHealthProviderSkips,
    pub created_at: DateTime<Utc>,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthCsvResolution {
    pub valid_rows: usize,
    pub skipped_rows: usize,
    pub duplicate_rows: usize,
    pub malformed_rows: usize,
}

#[derive(Clone, Debug, Default, Serialize)]
pub struct MarketHealthProviderSkips {
    pub finviz: Vec<MarketHealthProviderSkip>,
    pub yahoo: Vec<MarketHealthProviderSkip>,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthProviderSkip {
    pub symbol: TickerSymbol,
    pub message: String,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthTabResponse {
    pub tab: String,
    pub latest_session: NaiveDate,
    pub charts: Vec<MarketHealthChart>,
    pub leaders: Vec<MarketHealthLeader>,
    pub healthy_leaders: Vec<MarketHealthLeader>,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthChart {
    pub title: String,
    pub percent: bool,
    pub series: Vec<MarketHealthSeries>,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthSeries {
    pub name: String,
    pub points: Vec<MarketHealthPoint>,
    pub summary: MarketHealthSummary,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthPoint {
    pub date: NaiveDate,
    pub value: f64,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthSummary {
    pub current: Option<f64>,
    pub change_5d: Option<f64>,
    pub change_20d: Option<f64>,
}

#[derive(Clone, Debug, Serialize)]
pub struct MarketHealthLeader {
    pub symbol: TickerSymbol,
    pub percentile: f64,
    pub sector: Option<String>,
    pub sector_industry_keys: Vec<String>,
    pub industry_key: Option<String>,
    pub industry_group: Option<String>,
}
